### PR TITLE
[Embedder API] Add next frame callback

### PR DESCRIPTION
**Pull request by loic-sharma**
Adds an embedder API to be notified when the next frame has been drawn. In the future, this will be used by the Windows embedder to hide the window until the first frame has been drawn. 

Part of https://github.com/flutter/flutter/issues/41980

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/embedder/embedder.cc [hunk 1]
```diff
@@ -2683,6 +2683,36 @@ FlutterEngineResult FlutterEngineScheduleFrame(FLUTTER_API_SYMBOL(FlutterEngine)
...
+
+  if (callback == nullptr) {
+    return LOG_EMBEDDER_ERROR(kInvalidArguments,
+                              "Next frame callback was null.");
+  }
+
+  flutter::EmbedderEngine* embedder_engine =
+      reinterpret_cast<flutter::EmbedderEngine*>(engine);
+
+  fml::WeakPtr<flutter::PlatformView> weakPlatformView =
```
Comment (chinmaygarde) [inline shell/platform/embedder/embedder.cc:2702]: Nit: `weak_platform_view`

File: shell/platform/embedder/embedder.h [hunk 2]
```diff
@@ -2510,6 +2510,25 @@ FLUTTER_EXPORT
...
+///
+/// @param[in]  engine     A running engine instance.
+/// @param[in]  callback   The callback to execute.
+/// @param[in]  user_data  A baton passed by the engine to the callback. This
+///                        baton is not interpreted by the engine in any way.
+///
+/// @return     The result of the call.
+///
+FLUTTER_EXPORT
+FlutterEngineResult FlutterEngineSetNextFrameCallback(
```
Comment (chinmaygarde) [inline shell/platform/embedder/embedder.h:2527]: Documentation: Clarify that FlutterEngineSetNextFrameCallback must be made on the platform thread (the thread on which the FlutterEngineRun call was made).

Comment (chinmaygarde) [reply shell/platform/embedder/embedder.h:2527]: Documentation: Clarify if the callback is invoked repeatedly or only for one next frame.

File: shell/platform/embedder/embedder.h [hunk 3]
```diff
@@ -2510,6 +2510,25 @@ FLUTTER_EXPORT
 FlutterEngineResult FlutterEngineScheduleFrame(FLUTTER_API_SYMBOL(FlutterEngine)
                                                    engine);
 
+//------------------------------------------------------------------------------
+/// @brief      Schedule a callback to be called after the next frame is drawn.
+///             This callback is made on an internal engine managed thread and
```
Comment (chinmaygarde) [inline shell/platform/embedder/embedder.h:2515]: (or an embedder managed thread if custom task runners are specified)

**Review by chinmaygarde — APPROVED** [review]
LGTM with nits and documentation suggestions.

File: shell/platform/embedder/embedder.h (hunk 3 above)
Comment (loic-sharma) [reply shell/platform/embedder/embedder.h:2515]: Ah, perhaps my original comment is incorrect. This runs on the raster thread, which is created by the embedder (and thus also managed by the embedder?). Should I just say it runs on the raster thread?

File: shell/platform/embedder/embedder.cc (hunk 1 above)
Comment (loic-sharma) [reply shell/platform/embedder/embedder.cc:2702]: Nice catch, this was a copy paste mistake from iOS code.

File: shell/platform/embedder/embedder.h (hunk 2 above)
Comment (loic-sharma) [reply shell/platform/embedder/embedder.h:2527]: Ah, I was under the impression that all embedder APIs had to be called from the platform thread. Will add, thanks!

File: shell/platform/embedder/embedder.cc (hunk 1 above)
Comment (loic-sharma) [reply shell/platform/embedder/embedder.cc]: Fixed

File: shell/platform/embedder/embedder.h (hunk 2 above)
Comment (loic-sharma) [reply shell/platform/embedder/embedder.h:2528]: Added

File: shell/platform/embedder/embedder.h (hunk 3 above)
Comment (loic-sharma) [reply shell/platform/embedder/embedder.h]: I switched this to say raster thread. I'll leave this conversation open for now, let me know if you'd prefer I go with your original message!

**Review by cbracken — APPROVED** [review]
Nice!

LGTM stamp from a Japanese personal seal

**Pull request merged**